### ISSUE TITLE
Change Guards to Classes

### ISSUE THREAD

**Issue by MikeRyan52** [ngrx/router#78]
The offline compiler statically analyzes services and provide statements to codegen DI as an optimization. We want ngrx/router to be compatible with the offline compiler, so we have decided to change guards to be classes. Here's what they will look like:

```ts
@Injectable()
export class AuthGuard implements Guard {
  constructor(private http: Http) { }

  check(transition: Transition) {
    return this.http.get('/auth')
      .map(res => true)
      .catch(() => Observable.of(false));
  }
}
```

The `check()` method receives a transition object whose interface looks like:
```ts
interface Transition {
  routeParams: Params;
  queryParams: Params;
  route: Route;
  locationChange: LocationChange;
}
```

Installing a guard will be the same:
```ts
const routes: Routes = [
  {
    path: '/admin',
    guards: [ AuthGuard ]
  }
];
```

**Comment by MikeRyan52** [comment]
cc @laurelnaiad @nathasm @littleStudent @fxck 

Would love to hear your comments on this change.

**Comment by laurelnaiad** [comment]
How would the Transition interface relate to the TraversalCandidate interface from the traversal middleware? Would the latter by modified to be copacetic with the former? 

I do see queryParams in the former but not the latter. Does that imply that queryParams would be part of route traversal/candidate matching?

**Comment by littleStudent** [comment]
Looks good to me, functionality seems to stay the same.
Only one question. Why do you prefer to use the `Transition` interface instead of passing in the 4 properties as single parameters to check()? Didnt the previous version work like that?

**Comment by frederikschubert** [comment]
I think it is often a better coding style to have parameter objects than long parameter lists (more than 3?) for functions. For example this is one refactoring that Martin fowler suggests in his book on Refactoring.

**Comment by MikeRyan52** [comment]
@littleStudent @frederikschubert Yeah, we don't want our users to have to memorize positional arguments and it limits our flexibility to provide more information with future revisions to the API.

@laurelnaiad You know our codebase better than us :) Instead of `Transition`, we'll be updating `TraversalCandidate` to include the missing information and will be providing an instance of this interface to guards.

**Comment by MikeRyan52** [comment]
Closed and released in v0.4.0
